sig-robotics/examples/Definition_of_cloud_robot_CRD_and_Operator/internal/controller/robotsync_controller.go

```diff
@@ -0,0 +1,279 @@
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the RobotSync object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.4/pkg/reconcile
+func (r *RobotSyncReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	glog.Println("Start RobotSync Reconcile")
+
+	// TODO(user): your logic here
+	// GET RobotSync
+	// robotSync := &cloudrobotkubeedgev1beta1.RobotSync{}
+	// if err := r.Get(ctx, req.NamespacedName, robotSync); err != nil {
+	// 	if errors.IsNotFound(err) {
+	// 		// resource not found, it might be deleted
+	// 		return ctrl.Result{}, nil
+	// 	}
+	// 	// Error fetching RobotSync resource
+	// 	logger.Error(err, "Failed to GET robotSync")
+	// 	return ctrl.Result{}, err
+	// }
+	robotSyncList := &cloudrobotkubeedgev1beta1.RobotSyncList{}
+	if err := r.List(ctx, robotSyncList); err != nil {
+		glog.Println("Failed to list the robotSync", err)
+		return ctrl.Result{}, err
+	}
+	robotSync := &robotSyncList.Items[0]
+
+	// GET NODE
+	node := &corev1.Node{}
+	if err := r.Get(ctx, req.NamespacedName, node); err != nil {
+		if errors.IsNotFound(err) {
+			// resource not found, it might be deleted
+			return ctrl.Result{}, nil
+		}
+		// Error fetching Node resource
+		logger.Error(err, "Failed to GET node")
+		return ctrl.Result{}, err
+	}
+
+	// ADD & UPDATE
+	annotations := node.GetAnnotations()
+	nodeName := node.GetName()
+
+	// 判断已注册列表里有没有这个 robotId，防止重复创建 robot
+	var robotId uint
+	var ok = false
+	if robotId_str := node.Labels["robotId"]; robotId_str != "" {
+		// 如果已经注册（node 中的 label 包含 robotId 且不为空）就获取这个 robotId
+		robotId_int, err := strconv.Atoi(robotId_str)
+		if err != nil {
+			glog.Println("Failed to convert robotId", err)
+			return ctrl.Result{}, err
+		} else {
+			robotId = uint(robotId_int)
+			ok = true // 标识已注册
+		}
+	}
+	// 如果引发 reconcile 的 node 已经注册（有 robotId 标签）不管（更新心跳由robot controller 接受ws消息来更新）
+	// 如果引发 reconcile 的 node 未注册（没有 robotId 标签）那么就只创建 robot，更新心跳由robot controller 接受ws消息来更新
+	if !ok || !ContainsElement(robotSync.Status.RegistedRobots, robotId) {
+		// 未注册情况
+		// Create Robot resource
+		robot, err := r.applyRobot(ctx, annotations)
+		if err != nil {
+			glog.Println(err, "Failed to create Robot resource")
+			return ctrl.Result{}, err
+		}
+		// Add to Registed Robot list + add robotId into node labels
+		err = r.updateAllRobotSync(ctx, node, robotSync, robot, nodeName, annotations)
+		if err != nil {
+			glog.Println("Failed to update RobotSync resource", err)
+			return ctrl.Result{}, err
+		}
+	}
+
+	glog.Println("finish RobotSync Reconcile")
+
+	return ctrl.Result{}, nil
+}
+
+// Done
+func (r *RobotSyncReconciler) handleNodeDelete(ctx context.Context, node *corev1.Node) error {
```

Comment:
The `handleNodeDelete` function is responsible for handling the deletion of a `Node` resource. It correctly fetches the `Robot` resource associated with the node and removes it from the registered list. However, there is no error handling for the `Delete` call. Consider adding error handling to ensure that any issues with deleting the `Robot` resource are properly handled.
```diff
-	r.Delete(ctx, robot)
+	if err := r.Delete(ctx, robot); err != nil {
+		glog.Println("Failed to delete robot", err)
+		return err
+	}
```